### PR TITLE
Safer way to handle unlock command of cli_wallet #1171

### PR DESCRIPTION
**Pull request by cogutvalera**
PR for #1171 

bitshares/bitshares-fc#82

### CODE REVIEW

**Comment by cogutvalera** [comment]
this PR https://github.com/troglobit/editline/pull/21 was merged already.

Guys check please other 2 PRs when you will have time for it:
1. current #1382 
2. https://github.com/bitshares/bitshares-fc/pull/82

Thanks !

**Comment by cogutvalera** [comment]
fixed conflicts with latest `develop` branch

**Comment by cogutvalera** [comment]
will bump FC now after merged PR https://github.com/bitshares/bitshares-fc/pull/82

**Comment by cogutvalera** [comment]
need to be rebased after merging PR #1405

**Comment by cogutvalera** [comment]
rebased already

File: programs/cli_wallet/main.cpp [hunk 1]
```diff
@@ -72,6 +72,9 @@ int main( int argc, char** argv )
 {
    try {
 
+      const char* regex_secret_name = "regex-secret";
+      const string regex_secret = "^(unlock|set_password)\\s.*";
```
Comment (pmconrad) [inline programs/cli_wallet/main.cpp]: Must allow ``\s*`` at the start of the line.

**Comment by pmconrad** [comment]
Seems to work as designed

**BUT**

this solution is unusable IMO. Once the line matches the regexp there is no visual feedback - no cursor movement, no placeholder for keys pressed etc..

**Comment by cogutvalera** [comment]
I wanted to add placeholder '*' but made without it for security reasons for example as git works in command line with passwords

File: programs/cli_wallet/main.cpp (hunk 1 above)
Comment (cogutvalera) [reply programs/cli_wallet/main.cpp]: Done ! Thanks !

**Comment by pmconrad** [comment]
I'm aware of two ways to enter passwords (well, three if you count cleartext):
* a separate prompt for only the password, with no visual feedback
* visual feedback where password characters are replaced with asterisk or bullet points

I've never seen a normal editable command line that suddenly stops all visual feedback while still accepting (and expecting) input. A normal user will think the client is frozen, and will try to kill it, and then complain here.

**Comment by cogutvalera** [comment]
ok ! I will improve it !
> visual feedback where password characters are replaced with asterisk or bullet points
or whitespaces also possible (just cursor feedback) ? Right ?

**Comment by cogutvalera** [comment]
I've added new PR in `editline` project https://github.com/troglobit/editline/pull/22 for visual feedback while typing secret information

**Comment by pmconrad** [comment]
> or whitespaces also possible (just cursor feedback) ? Right ?

I think something other would be preferrable. Try to imagine how you as an unsuspecting user would react when observing that behaviour for the first time. I bet you'd find it confusing.

**Comment by cogutvalera** [comment]
@pmconrad I mean cursor movement

> Seems to work as designed
> 
> **BUT**
> 
> this solution is unusable IMO. Once the line matches the regexp there is no visual feedback - no cursor movement, no placeholder for keys pressed etc..

As you told cursor movement is already done in new PR for `editline` here https://github.com/troglobit/editline/pull/22

**Comment by cogutvalera** [comment]
@pmconrad did you have chance to check locally how works my new PR in `editline` with visual feedback by cursor as you told earlier to me ? Or didn't you have time to check it ? If yes, then later when having time to check it tell me please if it is good enough or not ?

Thanks !

JFYI @troglobit commented here https://github.com/troglobit/editline/pull/22#issuecomment-435353108 in my new PR with visual feedback by cursor. I'm thinking what should I answer to him cause my answer depends on your review. Thanks for your time and efforts !

**Comment by cogutvalera** [comment]
rebased due to conflicts

**Comment by cogutvalera** [comment]
Guys what should we do with this PR ? Look please when you will have time + new PR in `FC` library

https://github.com/bitshares/bitshares-fc/pull/97

**Comment by troglobit** [comment]
@cogutvalera I'm still open to a PR that adds visual feedback when in `el_no_echo=1` mode.  E.g., an `el_no_echo_indicator="*"` or similar would be useful.

**Comment by pmconrad** [comment]
You can remove the change in `CMakeLists.txt` - the dependency is inherited from `fc`.

I don't think it makes much sense to have a command line option for the regex. Better remove it.

**Comment by cogutvalera** [comment]
> You can remove the change in `CMakeLists.txt` - the dependency is inherited from `fc`.
> 
> I don't think it makes much sense to have a command line option for the regex. Better remove it.

About command line option for the regex I did it before you told earlier that it can be more flexible if there will be possibility to configure it for any desired data like brain key for example (look here your comment https://github.com/bitshares/bitshares-core/issues/1171#issuecomment-429518970). I see next 3 possible ways for configuring this mechanism:

1. Configure by command line argument
2. Configure by config file
3. Configure by both: command line arg & config file

Which one do you prefer @pmconrad ? Guys what do you think too ? What is the best way ?

**Comment by pmconrad** [comment]
What I meant with "configurable by the calling application" is that `fc::rpc::cli` shouldn't have this hardcoded, but that the application that uses `fc::rpc::cli` should define the regex. I didn't mean "configurable by the user".

IMO neither cli option nor config file option make sense. Definition via `static const` in `main.cpp` is sufficient.

**Comment by pmconrad** [comment]
Merged https://github.com/bitshares/bitshares-fc/pull/97 . Please rebase on latest  develop and bump fc to latest master.

File: programs/cli_wallet/main.cpp [hunk 2]
```diff
@@ -86,6 +89,7 @@ int main( int argc, char** argv )
          ("wallet-file,w", bpo::value<string>()->implicit_value("wallet.json"), "wallet to load")
          ("chain-id", bpo::value<string>(), "chain ID to connect to")
          ("suggest-brain-key", "Suggest a safe brain key to use for creating your account")
+         (regex_secret_name, bpo::value<string>()->implicit_value(regex_secret), "Regex to hide secret information")
```
Comment (abitmore) [inline programs/cli_wallet/main.cpp]: I don't think we need a parameter.

File: programs/cli_wallet/main.cpp [hunk 3]
```diff
@@ -72,6 +72,9 @@ int main( int argc, char** argv )
 {
    try {
 
+      const char* regex_secret_name = "regex-secret";
+      const string regex_secret = "^\\s*(unlock|set_password)\\s*";
```
Comment (abitmore) [inline programs/cli_wallet/main.cpp]: Why not `^\\s*(unlock|set_password)\\s*$`?

IMHO good UX is:
* if the user typed `unlock<ENTER>`, we prompt for a password;
* if the user typed `unlock SOMETHING<ENTER>`, we use `SOMETHING` as password.

Thoughts?

File: CMakeLists.txt [hunk 4]
```diff
@@ -53,7 +53,8 @@ LIST(APPEND BOOST_COMPONENTS thread
                              chrono
                              unit_test_framework
                              context
-                             locale)
+                             locale
+                             regex)
```
Comment (abitmore) [inline CMakeLists.txt]: I don't think this change is needed.

File: programs/cli_wallet/main.cpp (hunk 3 above)
Comment (cogutvalera) [reply programs/cli_wallet/main.cpp]: It works now right this way if I remember it correctly ! You should check it to see ...

Comment (abitmore) [reply programs/cli_wallet/main.cpp]: Yes, it's current behavior. Thanks.

I guess the `^` is not needed. Not a big deal though.

File: CMakeLists.txt (hunk 4 above)
Comment (cogutvalera) [reply CMakeLists.txt]: done

File: programs/cli_wallet/main.cpp (hunk 3 above)
Comment (cogutvalera) [reply programs/cli_wallet/main.cpp]: done

File: programs/cli_wallet/main.cpp (hunk 2 above)
Comment (cogutvalera) [reply programs/cli_wallet/main.cpp]: done

File: programs/cli_wallet/main.cpp [hunk 5]
```diff
@@ -257,6 +256,9 @@ int main( int argc, char** argv )
       if( !options.count( "daemon" ) )
       {
          auto wallet_cli = std::make_shared<fc::rpc::cli>( GRAPHENE_MAX_NESTED_OBJECTS );
+         
+         wallet_cli->set_regex_secret("\\s*(unlock|set_password)\\s*$");
```
Comment (abitmore) [inline programs/cli_wallet/main.cpp]: I think it's easier to understand if either have both `^` and `$`, or have none of them.

Comment (cogutvalera) [reply programs/cli_wallet/main.cpp]: done

**Comment by pmconrad** [comment]
Please squash your changes, we don't need 5 commits in the history for a one-liner.

**Comment by cogutvalera** [comment]
> Please squash your changes, we don't need 5 commits in the history for a one-liner.

ok sure, not a problem at all !

**Comment by cogutvalera** [comment]
Nice that it is already merged ! Thank you !

**Pull request merged**